jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IO.java

```diff
@@ -224,36 +224,110 @@ public static void copyDir(File from, File to) throws IOException
+            throw new IllegalArgumentException("Source is not a directory: " + srcDir);
+        if (!Files.isDirectory(Objects.requireNonNull(destDir)))
+            throw new IllegalArgumentException("Dest is not a directory: " + destDir);
+
+        try (Stream<Path> sourceStream = Files.walk(srcDir))
+        {
+            Iterator<Path> iterFiles = sourceStream
+                .filter(Files::isRegularFile)
+                .iterator();
+            while (iterFiles.hasNext())
+            {
+                Path sourceFile = iterFiles.next();
+                Path relative = srcDir.relativize(sourceFile);
+                Path destFile = resolvePath(destDir, relative);
+                if (!Files.exists(destFile.getParent()))
+                    Files.createDirectories(destFile.getParent());
+                copyFile(sourceFile, destFile);
+            }
+        }
+    }
+
+    /**
+     * Copy the contents of a source directory to destination directory.
+     *
+     * @param srcDir the source directory
+     * @param destDir the destination directory
+     * @param copyOptions the options to use on the {@link Files#copy(Path, Path, CopyOption...)} commands.
+     * @throws IOException if unable to copy the file
+     */
     public static void copyDir(Path srcDir, Path destDir, CopyOption... copyOptions) throws IOException
     {
         if (!Files.isDirectory(Objects.requireNonNull(srcDir)))
             throw new IllegalArgumentException("Source is not a directory: " + srcDir);
         if (!Files.isDirectory(Objects.requireNonNull(destDir)))
             throw new IllegalArgumentException("Dest is not a directory: " + destDir);
 
-        try (Stream<Path> sourceStream = Files.walk(srcDir, 20))
+        try (Stream<Path> sourceStream = Files.walk(srcDir))
         {
             Iterator<Path> iterFiles = sourceStream
                 .filter(Files::isRegularFile)
                 .iterator();
             while (iterFiles.hasNext())
             {
                 Path sourceFile = iterFiles.next();
-                URI relativeSrc = srcDir.toUri().relativize(sourceFile.toUri());
-                Path destFile = destDir.resolve(relativeSrc.toASCIIString());
+                Path relative = srcDir.relativize(sourceFile);
+                Path destFile = resolvePath(destDir, relative);
                 if (!Files.exists(destFile.getParent()))
                     Files.createDirectories(destFile.getParent());
                 Files.copy(sourceFile, destFile, copyOptions);
             }
         }
     }
 
+    /**
+     * Perform a resolve of a path against a relative path in a
+     * way that is smart about {@link java.nio.file.FileSystem} differences.
+     *
+     * @param basePath the base Path
+     * @param relative the Path to resolve against base Path
+     * @return the new Path object relative to the base Path
+     */
+    public static Path resolvePath(Path basePath, Path relative)
```

Comment:
Was this piece of code a need to be able to get rid of the `Path` -> `URI` -> `Path` conversion? If so, you should probably explain that in the javadoc, and maybe give it a better name?

Reply:
Yes, this is the combination of the old `Resource.resolveDifferentFileSystem(Path, Path)` along with the FileSystem check for the two paths.
Javadoc improvements, yes, will do!

The goal for this method is to be a FileSystem neutral way to use `Path.resolve(Path)`, so I left it as simply `resolvePath(Path, Path)` as that made the most sense to me.
Got any opinions on names?

Reply:
`resolveAcrossFileSystems()`?

Reply:
The "across FileSystems" isn't quite right.
It resolves the "basePath" against a relative path, ignoring FileSystem behaviors.

Reply:
I updated the javadoc.
If you still are not happy with the name, perhaps the new javadoc make it more clear of it's purpose.

Reply:
Okay, javadoc makes it clear enough. If you just added the extra check `if (relative.isAbsolute()) throw new IllegalArgumentException()` this method would be perfect.